### DISCUSSION TITLE
2.6?

### DISCUSSION THREAD

**Discussion by ObviousInRetrospect** [SpenceKonde/megaTinyCore#795] [Q&A]
Looks like a 2.6 release was cut. Is it expected to work with the old toolchain or should we wait for the new one?

Also, assuming that head of tree is now clean, any thoughts about committing potentially breaking and/or partial changes to a dev branch and then merging over to master to keep head of tree usable if more bleeding edge than the releases? Would make testing pull requests cleaner.

Thanks!

**Comment by ObviousInRetrospect** [comment]
Also might be worth noting in the release notes that Wire as a slave now works reliably with sleep.

**Comment by SpenceKonde** [comment]
Well, right now we're at a standstill because I can't host the x64 linux toolchain file. 

Github is erroring when I try to upload it either as an attachment or into gh-pages, and I can't host it the way I was hosting the other files because the bandwidth costs (which are negligible normally) become prohibitive when the CI tests are hammering it. We really really want the toolchain to be hosted somewhere on github, because the downloads by the CI system are multiple orders of magnitude greater than all other downloads combined. But github says the file is too large for github pages; In the past I have attached it to a release, but now I get a non-specific error message informatively telling me that something went wrong. 

Each CI run involves at least 72 installations of the core, in this case it comes to about 25 cents per CI run. But I think it may actually be doing the download for every call to CompileExamples - in which case it's around $2.30 per CI run (based on the cost of bandwidth from AWS). I've been hosting the other platforms on AWS S3, and the costs are fine when we only have to supply the package for normal people - about a third of a cent per installation every few versions is not problematic. But $2.30 or even $0.25 per run of the compile tests really adds up, because that happens with almost every commit that touches code; obviously everyone is better off if the transfer stays within github for the CI tests. So I'm not quite sure what to do.

**Reply by ObviousInRetrospect** [reply]
Uh any chance you can just stick it in a new repo and download it via the raw path (and/or a release in a new repo to try to roll the dice again on the error).

How big was the last version, how big is this version?

Any chance this is useful? https://docs.github.com/en/github-ae@latest/repositories/working-with-files/managing-large-files/about-git-large-file-storage

**Comment by SpenceKonde** [comment]
Okay, the limit is 50MB through the client, so x64 linux is up. 

But now we have another roadblock that cannot be surmounted until monday: I can't log into AWS to upload the rest of the files and I feel nervous about pushing my luck to host it all on github (plus, there's no way to get any analytics on it - granted AWS's analytics are about as clear as mud - , because my phone's charging port has failed (falsely reports that there is moisture in the charging port) and is at 0% , so I can't get the 2FA message so I can't log in or upload anything. It sounds like wireless charging will work, and I've got a wireless charger arriving tomorrow - but it's coming to my HQ in cambridge, and I'm not going to get back there until monday.

**Reply by ObviousInRetrospect** [reply]
I don't think hosting all the files involved in installing a release on GitHub is (even close to) abusive after reading this: https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases 

re AWS: after you get back in you might want to create an IAM user (and probably a key) with just S3 write permissions and without MFA - no real reason to use console login to upload artifacts use the CLI. If you have such a key already you probably don't need your MFA.

**Reply by ObviousInRetrospect** [reply]
and actually from the perspective of manual install directions download the tools from the release is way less ugly than go get the json and find the path to the tools in s3.

**Reply by SpenceKonde** [reply]
You seem to be under the mistaken belief that I can make  another user (i can'r,.remember, I can't log in) or that tht indicidual would have hel[ed becasue I think there's a second round of 2Fa tp cjhange settings .ikler tjnat
 would have helped. Because how would  have logged in to create that user? By entering m 2FA code. Maybe that user would also need 2FA. Andante level of security would I fall back to?! A stupid pw? Which no human can remember? I think about half the times I am trying to log in, I get to the point where I get punished (extra long delay, multiple CAPTCAs consecutively, and so-on But I don't know the password? How the hall would I remember I passcode I made *ONCE* many months ago. And many of these places have a blacklist of passwords a mile lend everywhere has different complexity rules. now with the 2FA. Amd you can't use a password manager,because thy keep getting hacked. So it's basically if the passwords isn't saved in the browser and 2Fa device isn't handy. Either you have iit written down easy reach, you slog through pw recovery every time your browser frogets the username and passsowrd/ 

A, I unusual in this regard?

**Reply by ObviousInRetrospect** [reply]
Sorry, the suggestion to create a less privileged IAM user was after you got back in so you could avoid the problem in the future.  And yeah it would need a password that was strong enough to not need MFA. Nonsensical passwords like $imple0penDrainAnalogAdapter + a paper backup or password safe are actually pretty secure (well that one isn't ... anymore).

But really the better part of that suggestion was to check whether you had an access key lying around in which case you don't need MFA. Access keys allow API/CLI access using a far more secure hmac-sha256 protocol and a key that is too large to guess instead of a password, thus MFA isn't needed. https://awscli.amazonaws.com/v2/documentation/api/latest/reference/s3/index.html - just don't check the key into GitHub. If you don't, you should probably create one. In general I would again create an IAM user with appropriate permissions (S3 not EC2) and a key. And yes unless you already have one, after you get back in.

**Reply by SpenceKonde** [reply]
*admits that he would have no idea how to use such an access key*

**Reply by ObviousInRetrospect** [reply]
install the CLI I linked.  
```
python -m pip install awscli 
aws configure
```

Will prompt you for the key. (either in the system python or a virtualenv work) See https://github.com/aws/aws-cli#getting-started for other configuration options. 

But before you run random commands that ask for credentials for your aws account because someone told you to on a discussion board, you should also confirm that https://awscli.amazonaws.com/v2/documentation/api/latest/reference/s3/index.html links to https://github.com/aws/aws-cli#getting-started  and that GitHub repo gives installation directions that match what I wrote above :-).

You can then run things like

```
aws s3 ls
aws s3 cp foo.tar.gz s3://mybucket/foo.tar.gz
```

Don't check the key into GitHub. Keep it safe.

**Comment by SpenceKonde** [comment]
2.6.0 is now out in board manager, or it should be.

**Reply by ObviousInRetrospect** [reply]
Install is failing on a Mac.

Tool DxCore:avrdude@6.3.0-arduino17or18 already installed
Tool arduino:arduinoOTA@1.3.0 already installed
Tool megaTinyCore:python3@3.7.2-post1 already installed
Downloading packages
DxCore:avr-gcc@7.3.0-atmel3.6.1-azduino5
Failed to install platform: megaTinyCore:megaavr.
Error: 13 INTERNAL: Error downloading tool DxCore:avr-gcc@7.3.0-atmel3.6.1-azduino5: Server responded with: 403 Forbidden

(1.8.19 also fails with a non-copyable error)
![image](https://user-images.githubusercontent.com/42229858/190066667-016ae91d-d379-402c-bd69-0dcf1f03db62.png)

**Reply by ObviousInRetrospect** [reply]
yep

https://spencekondetoolchains.s3.amazonaws.com/avr-gcc-7.3.0-atmel3.6.1-azduino5-x86_64-apple-darwin14.tar.bz2 (pulled from the json):

```
<Error>
<Code>AccessDenied</Code>
<Message>Access Denied</Message>
<RequestId>redacted</RequestId>
<HostId>
9PjvU4ijAeiriKyyFBr8oPmdb+xsIKFPnwQtop+Gh/l5zV5N/llKZA5KUJ5IP40Dj5Wj4Lf/hLc=
</HostId>
</Error>
```

**Comment by SpenceKonde** [comment]
This is an issue, not a discussion. 

Fucking S3. In the promotional material they make it sound like it's great for hosting large binaries for public consumption. 

But they put a billion traps in there, terrified that a user will unintentionally share confidental info and sue them or something.

**Reply by ObviousInRetrospect** [reply]
verified all the below azduino5s are access denied
```
<Contents>
<Key>
avr-gcc-7.3.0-atmel3.6.1-azduino5-aarch64-pc-linux-gnu.tar.bz2
</Key>
<LastModified>2022-09-14T00:01:20.000Z</LastModified>
<ETag>"5ebc9d23d466dbd23b02df0cdfc4eee4-3"</ETag>
<Size>38084721</Size>
<StorageClass>STANDARD</StorageClass>
</Contents>
<Contents>
<Key>
avr-gcc-7.3.0-atmel3.6.1-azduino5-arm-linux-gnueabihf.tar.bz2
</Key>
<LastModified>2022-09-14T00:01:20.000Z</LastModified>
<ETag>"cd843abf00eb5fa90b4446444f6db4b5-3"</ETag>
<Size>34518741</Size>
<StorageClass>STANDARD</StorageClass>
</Contents>
<Contents>
<Key>
avr-gcc-7.3.0-atmel3.6.1-azduino5-i686-pc-linux-gnu.tar.bz2
</Key>
<LastModified>2022-09-14T00:01:20.000Z</LastModified>
<ETag>"9ccf3dccc7b6a03fd3a4393fd0a7718a-3"</ETag>
<Size>37178887</Size>
<StorageClass>STANDARD</StorageClass>
</Contents>
<Contents>
<Key>
avr-gcc-7.3.0-atmel3.6.1-azduino5-i686-w64-mingw32.tar.bz2
</Key>
<LastModified>2022-09-14T00:01:20.000Z</LastModified>
<ETag>"f7b0e90d9de9d4968fadef59d27bfc1b-3"</ETag>
<Size>44772652</Size>
<StorageClass>STANDARD</StorageClass>
</Contents>
<Contents>
<Key>
avr-gcc-7.3.0-atmel3.6.1-azduino5-x86_64-apple-darwin14.tar.bz2
</Key>
<LastModified>2022-09-14T00:01:20.000Z</LastModified>
<ETag>"3eb0a2ea38fc566e5556a1dd1b9cab48-3"</ETag>
<Size>37716679</Size>
<StorageClass>STANDARD</StorageClass>
</Contents>
<Contents>
<Key>
avr-gcc-7.3.0-atmel3.6.1-azduino5-x86_64-pc-linux-gnu.tar.bz2
</Key>
<LastModified>2022-09-14T00:10:04.000Z</LastModified>
<ETag>"51af88e5e2817f9f244b3e3c96f91e78-3"</ETag>
<Size>37730911</Size>
<StorageClass>STANDARD</StorageClass>
</Contents>
```

**Reply by SpenceKonde** [reply]
They should work now.

**Reply by ObviousInRetrospect** [reply]
I can confirm the Mac one works

**Reply by ObviousInRetrospect** [reply]
well it installs.
```
fork/exec /Users/user/Library/Arduino15/packages/DxCore/tools/avr-gcc/7.3.0-atmel3.6.1-azduino5/bin/avr-g++: no such file or directory

Compilation error: fork/exec /Users/user/Library/Arduino15/packages/DxCore/tools/avr-gcc/7.3.0-atmel3.6.1-azduino5/bin/avr-g++: no such file or directory
```

1.x repro:
```

fork/exec /Users/user/Library/Arduino15/packages/DxCore/tools/avr-gcc/7.3.0-atmel3.6.1-azduino5/bin/avr-g++: no such file or directory
Error compiling for board ATtiny412/402/212/202.
```

**Reply by ObviousInRetrospect** [reply]
looks like there is an extra avr in the path:
 /Users/user/Library/Arduino15/packages/DxCore/tools/avr-gcc/7.3.0-atmel3.6.1-azduino5/bin doesn't exist but  /Users/user/Library/Arduino15/packages/DxCore/tools/avr-gcc/7.3.0-atmel3.6.1-azduino5/avr/bin does.

/Users/user/Library/Arduino15/packages/DxCore/tools/avr-gcc/7.3.0-atmel3.6.1-azduino5/ contains only "avr"

Since  /Users/user/Library/Arduino15/packages/DxCore/tools/avr-gcc/7.3.0-atmel3.6.1-azduino5/avr/avr exists I am assuming an extra avr got put in the tar

**Reply by ObviousInRetrospect** [reply]
so after getting past that by moving the files into the right path I hit compilation errors:

```
/Users/user/Library/Arduino15/packages/megaTinyCore/hardware/megaavr/2.6.0/cores/megatinycore/UART.cpp: In member function 'void HardwareSerial::begin(long unsigned int, uint16_t)':
/Users/user/Library/Arduino15/packages/megaTinyCore/hardware/megaavr/2.6.0/cores/megatinycore/UART.cpp:499:24: error: 'USART_RXMODE0_bm' was not declared in this scope
             ctrlb   |= USART_RXMODE0_bm;        // set the U2X bit in what will become CTRLB
                        ^~~~~~~~~~~~~~~~
/Users/user/Library/Arduino15/packages/megaTinyCore/hardware/megaavr/2.6.0/cores/megatinycore/UART.cpp:499:24: note: suggested alternative: 'USART_RXMODE_0_bm'
             ctrlb   |= USART_RXMODE0_bm;        // set the U2X bit in what will become CTRLB
                        ^~~~~~~~~~~~~~~~
                        USART_RXMODE_0_bm
/Users/user/Library/Arduino15/packages/megaTinyCore/hardware/megaavr/2.6.0/cores/megatinycore/UART.cpp:522:19: error: 'USART_RS485_bm' was not declared in this scope
       if (ctrla & USART_RS485_bm) {             // RS485 mode recorded here too... because we need to set
                   ^~~~~~~~~~~~~~
/Users/user/Library/Arduino15/packages/megaTinyCore/hardware/megaavr/2.6.0/cores/megatinycore/UART.cpp:522:19: note: suggested alternative: 'USART_RS485_gm'
       if (ctrla & USART_RS485_bm) {             // RS485 mode recorded here too... because we need to set
                   ^~~~~~~~~~~~~~
                   USART_RS485_gm
exit status 1
Error compiling for board ATtiny412/402/212/202.
```

ironically this is the rare sketch that doesn't use Serial at all.

**Reply by ObviousInRetrospect** [reply]
Got this on a sketch that included had a gratuitous #include <Comparator.h>
```
In file included from /Users/user/Documents/Arduino/soilmoist3224b/soilmoist3224b.ino:19:0:
/Users/user/Library/Arduino15/packages/megaTinyCore/hardware/megaavr/2.6.0/libraries/Comparator/src/Comparator.h:323:5: error: 'in_comparator' does not name a type; did you mean 'Comparator'?
     in_comparator::p::inputP_t     input_p        = in_comparator::p::in0;
     ^~~~~~~~~~~~~
     Comparator
/Users/user/Library/Arduino15/packages/megaTinyCore/hardware/megaavr/2.6.0/libraries/Comparator/src/Comparator.h:324:5: error: 'in_comparator' does not name a type; did you mean 'Comparator'?
     in_comparator::n::inputN_t     input_n        = in_comparator::n::in0;
     ^~~~~~~~~~~~~
     Comparator

exit status 1
Error compiling for board ATtiny3224/1624/824/424 w/Optiboot.
```

when I commented that out I got the USART error again (on a sketch that does use Serial)

**Reply by SpenceKonde** [reply]
Eeuuuuggghhhhhhhhhhhhhhhhh

Where is the compatibility fix that's supposed to be in core_devices to catch that? It's a 3000 line blob of auto-generated text, and it's not there (3 lines of preprocessor macro per macro that they renamed, and they renamed like a thousand variables. At least they had the decency to admit that it was a new major version

I have no idea what is going on with comparator.

**Comment by ObviousInRetrospect** [comment]
Cut 3 issues for the above but at least on Mac 2.6 seems to have some problems. Will do more troubleshooting tomorrow, have to be awake in too few hours.

**Comment by SpenceKonde** [comment]
I've pushed fixes to most of the other issues. 

The directory structure for the toolchain packages is one of the most aggressively idiotic things in the universe - it's on the first page of the idiotic ideas list! I have managed to fail to correctly pack the files on at least one platform every single time I have done a toolchain package. At least this time I consistently packed it wrong everywhere. 

There aren't many mere mortals to get an entry on that list of idiotic things; most of the list is bad physical ideas, like the stupid speed of light (ffs, seriously? But you put the stars *light years* apart, is this some sort of lab experiment where you're trying to keep each solar system isolated?)... and then there's that bullshit about entropy that, writ large, says that the universe can only get shittier, and the only way to make it less shitty in one place is by making another place shittier). Superconductivity only happening at temperatures too cold or pressures too high for life? C'mon, don't you tell me you don't want a superconductive nervous system - Of course you do! In amongst that is "tarball format ambiguous about top level directory, yet standard on all civilized planets"

Terrible ideas - we totally got shafted - the universe is a ripoff! If I was God's boss, I'd have fired him billions of years ago. In fact maybe that happened! That would explain why we don't seem to have anyone running the show up there... Maybe instead of St. Peter, when good people die they find themselves in a crowd of other good people and aliens, the pearly gates locked, and a "now hiring" sign hanging across them.... 

That's not a problem I have to worry about though, thankfully - we wicked folk know how to have a party anywhere and anywhen, even if the venue (in this case hell) gets shut down (surely you know the phrase "louder than hell"?). We've practiced for that. I can imagine why the devil would decamp for rural georgia and try his hand at the fiddle a while back: the whole joint probably got trashed by a bunch of dead sinners throwing a rager, and he just couldn't take it anymore and went on vacation while the repair demons got hell working again.... Knowing him, that wasn't *his* golden fiddle he was wagering either, probably snatched it from some dude who took it with him to the grave.

**Reply by SpenceKonde** [reply]
How do I tell tar to not include the containing directory? I kept very detailed notes this time: 
I was in a directory with one folder for each architecture, each containing an avr directory with the correct structure. I typed
```
tar -cf avr-gcc-7.3.0-atmel3.6.1-azduino5-aarch64-pc-linux-gnu.tar aarch64-pc-linux-gnu/avr
tar -cf avr-gcc-7.3.0-atmel3.6.1-azduino5-arm-linux-gnueabihf.tar arm-linux-gnueabihf/avr
tar -cf avr-gcc-7.3.0-atmel3.6.1-azduino5-i686-pc-linux-gnu.tar i686-pc-linux-gnu/avr
tar -cf avr-gcc-7.3.0-atmel3.6.1-azduino5-x86_64-apple-darwin14.tar x86_64-apple-darwin14/avr
tar -cf avr-gcc-7.3.0-atmel3.6.1-azduino5-x86_64-pc-linux-gnu.tar x86_64-pc-linux-gnu/avr
```
So I have told it I want to tar up the avr directory, why is it grabbing the containing one? This is so fucking stupid. 

There's gotta be a better way than
```
cd aarch64-pc-linux-gnu
tar -cf avr-gcc-7.3.0-atmel3.6.1-azduino5-aarch64-pc-linux-gnu.tar  avr
cd ..
cd arm-linux-gnueabihf
tar -cf avr-gcc-7.3.0-atmel3.6.1-azduino5-arm-linux-gnueabihf.tar  avr
cd ..
cd i686-pc-linux-gnu
tar -cf avr-gcc-7.3.0-atmel3.6.1-azduino5-i686-pc-linux-gnu.tar  avr
cd ..
cd x86_64-apple-darwin14
tar -cf avr-gcc-7.3.0-atmel3.6.1-azduino5-x86_64-apple-darwin14.tar  avr
cd ..
cd x86_64-pc-linux-gnu
tar -cf avr-gcc-7.3.0-atmel3.6.1-azduino5-x86_64-pc-linux-gnu.tar  avr
cd ..
```

**Reply by ObviousInRetrospect** [reply]
uh the way I would do it is:

```
cd x86_64-apple-darwin14/avr
tar -cf ../avr-gcc-7.3.0-atmel3.6.1-azduino5-x86_64-apple-darwin14.tar  .
```

some versions of tar support a -C which does basically that:
```
cd x86_64-apple-darwin14
tar -cf avr-gcc-7.3.0-atmel3.6.1-azduino5-x86_64-apple-darwin14.tar  -C avr .
```

**Comment by SpenceKonde** [comment]
Toolchain repacked and uploaded as Azduino6, json file updated for new toolchain and hosted copies of Azduino5 taken down so people can't install it. Description in board manager has been changed to say that it does not work (you must absolutely never ever ever remove anything from board manager json, yasee - then the IDE can't remove it! Yes, that's right, there's no record of what things it's installed stored separately from the latest downloaded version of the json. Did I get these archives right? (there's no core yet and I need sleep, but can you check that they have the right directory structure?

**Reply by ObviousInRetrospect** [reply]
not convinced:

the old one contained:
[edit: by old one I meant the azduino5 release not the working one]

x86_64-apple-darwin14/avr/*

the new one contains:
avr/*

based on the error message I had moved:
x86_64-apple-darwin14/avr/* -> x86_64-apple-darwin14/*
[in the files installed by board manager]

which worked.

I think you want to remove avr instead of  x86_64-apple-darwin14 ?

**Reply by ObviousInRetrospect** [reply]
well 6 has the same structure as 4a so you might have it right. I don't know how the installer works I was basing what I was saying on the deltas I had to apply

**Reply by ObviousInRetrospect** [reply]
Sorry, error, insufficient coffee. Unless you are trying to enable board manager installs before sleep (and why) this will probably go better after some sleep.

**Reply by ObviousInRetrospect** [reply]
```
% for i in avr-gcc-7.3.0-atmel3.6.1-azdu*.tar.bz2; do echo $i; tar tf $i | head -5; done
avr-gcc-7.3.0-atmel3.6.1-azduino4a-x86_64-apple-darwin14.tar.bz2
avr/
avr/avr/
avr/avr/bin/
avr/avr/bin/ar
avr/avr/bin/as

avr-gcc-7.3.0-atmel3.6.1-azduino5-x86_64-apple-darwin14.tar.bz2
x86_64-apple-darwin14/avr/
x86_64-apple-darwin14/avr/avr/
x86_64-apple-darwin14/avr/avr/bin/
x86_64-apple-darwin14/avr/avr/bin/ar
x86_64-apple-darwin14/avr/avr/bin/as

avr-gcc-7.3.0-atmel3.6.1-azduino6-x86_64-apple-darwin14.tar.bz2
avr/
avr/avr/
avr/avr/bin/
avr/avr/bin/ar
avr/avr/bin/as
```

**Reply by ObviousInRetrospect** [reply]
(translation: 6 is fine, sleep(_cpu))

**Comment by ObviousInRetrospect** [comment]
what is the current state believed to be?

I haven't been able to get anything to compile on head of tree. #807 makes things slightly more tolerable as a stopgap
